File: native/android/voice/sakha/SakhaTtsPlayer.kt
/**
 * Sakha TTS Player
 *
 * Sequential playback queue for the Sakha voice. Two backends:
 *  1. **Sarvam REST** (preferred for prose + verses)
 *     POST {backend}/api/voice-companion/synthesize → audio bytes (base64 in JSON
 *     or raw audio/* depending on tier). Uses MediaPlayer for playback.
 *  2. **Android system TTS** (fallback)
 *     Used when Sarvam fails *and* the language has a system voice. Never used
 *     for Sanskrit — system TTS will mangle it.
 *
 * The player is pause-aware: it consumes [PauseEvent]s from the parser and
 * either:
 *  - synthesises + plays the segment (Speak),
 *  - delays for the requested number of milliseconds (Pause), or
 *  - flushes everything and emits FILTER_FAIL handling (Filter).
 *
 * Concurrency model:
 *  - One coroutine pulls events off the inbound channel and enqueues
 *    Synthesise jobs.
 *  - One coroutine drains the synthesis queue, plays each clip back-to-back,
 *    and signals progress via [SakhaVoiceListener].
 *
 * This keeps synthesis pipelined ahead of playback (so the user hears one
 * sentence while the next is being synthesised) without ever overlapping
 * audio — Sakha's pacing depends on ordered, gapless playback.
 */

package com.mindvibe.kiaan.voice.sakha

import android.content.Context
import android.media.AudioAttributes
import android.media.MediaPlayer
import android.net.Uri
import android.speech.tts.TextToSpeech
import android.speech.tts.UtteranceProgressListener
import android.util.Base64
import android.util.Log
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeoutOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.File
import java.io.FileOutputStream
import java.util.UUID
import java.util.concurrent.TimeUnit

class SakhaTtsPlayer(
    private val context: Context,
    private val config: SakhaVoiceConfig,
    private val listener: SakhaVoiceListener,
) {

    companion object {
        private const val TAG = "SakhaTtsPlayer"
        private val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()
    }

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val httpClient = OkHttpClient.Builder()
        .connectTimeout(8, TimeUnit.SECONDS)
        .readTimeout(20, TimeUnit.SECONDS)
        .build()

    // Channels are recreated on every [start] — Kotlin Channel.close() is
    // terminal, so a session that stops and restarts (e.g. barge-in) needs
    // fresh ones or trySend will silently no-op on the second turn.
    private var incoming: Channel<PauseEvent> = Channel(Channel.UNLIMITED)
    private var playbackQueue: Channel<PlayJob> = Channel(Channel.UNLIMITED)
    private var ingestJob: Job? = null
    private var playbackJob: Job? = null

    private var fallbackTts: TextToSpeech? = null
    private var fallbackTtsReady = false

    private var currentPlayer: MediaPlayer? = null
    private var totalSpokenMs: Long = 0L
    private var pauseCount: Int = 0
    private var stopped: Boolean = false

    /**
     * Start a new playback session. Call [enqueue] to feed events and
     * [stop] to halt and free resources.
     */
    fun start() {
        stopped = false
        totalSpokenMs = 0L
        pauseCount = 0
        ensureFallbackTts()

        ingestJob?.cancel()
        playbackJob?.cancel()

        // Fresh channels for the new session.
        try { incoming.close() } catch (_: Exception) {}
        try { playbackQueue.close() } catch (_: Exception) {}
        incoming = Channel(Channel.UNLIMITED)
        playbackQueue = Channel(Channel.UNLIMITED)

        ingestJob = scope.launch {
            for (event in incoming) {
                if (stopped) break
                handleIncoming(event)
            }
            playbackQueue.close()
        }

        playbackJob = scope.launch {
            for (job in playbackQueue) {
                if (stopped) break
                playJob(job)
            }
        }
    }

    /** Feed a parser event into the player. Non-blocking. */
    fun enqueue(event: PauseEvent) {
        if (stopped) return
        incoming.trySend(event)
    }

    /** Signal that the SSE stream is finished — drain remaining buffered events. */
    fun finish() {
        incoming.close()
    }

    /** Stop immediately and release MediaPlayer + TTS resources. */
    fun stop() {
        stopped = true
        try { incoming.close() } catch (_: Exception) {}
        try { playbackQueue.close() } catch (_: Exception) {}
        currentPlayer?.let {
            try { it.stop() } catch (_: Exception) {}
            try { it.release() } catch (_: Exception) {}
        }
        currentPlayer = null
        ingestJob?.cancel()
        playbackJob?.cancel()
    }

    /** Permanently release. Call from onDestroy. */
    fun shutdown() {
        stop()
        try { fallbackTts?.shutdown() } catch (_: Exception) {}
        fallbackTts = null
        scope.cancel()
    }

    fun statsTotalSpokenMs(): Long = totalSpokenMs
    fun statsPauseCount(): Int = pauseCount

    // ------------------------------------------------------------------------
    // Pipeline
    // ------------------------------------------------------------------------

    private suspend fun handleIncoming(event: PauseEvent) {
        when (event) {
            is PauseEvent.Speak -> {
                // Defence-in-depth: persona guard rewrites in-place.
                val safeText = if (config.enablePersonaGuard) {
                    val result = SakhaPersonaGuard.softenInline(event.text)
                    if (result.text.isBlank()) {
                        // Phrase was excised entirely; nothing left to say.
                        return
                    }
                    result.text
                } else {
                    event.text
                }
                playbackQueue.trySend(PlayJob.Synthesise(safeText, event.isSanskrit))
            }
            is PauseEvent.Pause -> {
                playbackQueue.trySend(PlayJob.Pause(event.durationMs))
            }
            is PauseEvent.Filter -> {
                playbackQueue.trySend(PlayJob.Filter)
            }
        }
    }

    private suspend fun playJob(job: PlayJob) {
        when (job) {
            is PlayJob.Synthesise -> {
                val audioFile = withTimeoutOrNull(8_000L) { synthesise(job.text, job.isSanskrit) }
                if (audioFile != null) {
                    val ms = playFile(audioFile)
                    totalSpokenMs += ms
                    listener.onSpokenSegment(job.text, job.isSanskrit)
                    runCatching { audioFile.delete() }
                } else {
                    // Fallback to system TTS where possible (never for Sanskrit).
                    if (!job.isSanskrit && fallbackTtsReady) {
                        val ms = speakWithFallback(job.text)
                        totalSpokenMs += ms
                        listener.onSpokenSegment(job.text, false)
                    } else {
                        listener.onError(SakhaVoiceError.TtsError("synthesis-failed"))
                    }
                }
            }
            is PlayJob.Pause -> {
                pauseCount++
                listener.onPause(job.durationMs)
                delay(job.durationMs)
            }
            PlayJob.Filter -> {
                stop()
                listener.onFilterFail()
            }
        }
    }

    // ------------------------------------------------------------------------
    // Sarvam REST synthesis
    // ------------------------------------------------------------------------

    private suspend fun synthesise(text: String, isSanskrit: Boolean): File? {
        val url = "${config.backendBaseUrl.trimEnd('/')}${config.voiceCompanionPathPrefix}/synthesize"
        val voiceId = if (isSanskrit) config.sanskritVoiceId else config.sakhaVoiceId
        val body = JSONObject().apply {
            put("text", text)
            put("voice_id", voiceId)
            put("language", if (isSanskrit) "sa" else config.language.wire)
            put("speed", if (isSanskrit) 0.85 else 1.0)
            put("output_format", "mp3")
        }.toString()

        val token = config.authTokenProvider.invoke()
        val builder = Request.Builder()
            .url(url)
            .post(body.toRequestBody(JSON_MEDIA_TYPE))
            .header("X-Client", "kiaanverse-android-sakha-voice")
            .header("X-Voice-Mode", "sakha")
        token?.let { builder.header("Authorization", "Bearer $it") }

        return try {
            val response = httpClient.newCall(builder.build()).execute()
            response.use { resp ->
                if (!resp.isSuccessful) {
                    if (config.debugMode) Log.w(TAG, "Synth HTTP ${resp.code}")
                    return null
                }
                val contentType = resp.header("Content-Type", "audio/mpeg") ?: "audio/mpeg"
                val bytes: ByteArray = if (contentType.startsWith("application/json")) {
                    val json = JSONObject(resp.body?.string() ?: return null)
                    val b64 = json.optString("audio_base64").ifEmpty { json.optString("audio") }
                    if (b64.isNullOrEmpty()) return null
                    Base64.decode(b64, Base64.DEFAULT)
                } else {
                    resp.body?.bytes() ?: return null
                }
                writeTempAudio(bytes, contentType)
            }
        } catch (e: Exception) {
            if (config.debugMode) Log.w(TAG, "Synth failed: ${e.message}")
            null
        }
    }

    private fun writeTempAudio(bytes: ByteArray, contentType: String): File {
        val ext = when {
            contentType.contains("mpeg") -> "mp3"
            contentType.contains("wav") -> "wav"
            contentType.contains("ogg") -> "ogg"
            else -> "mp3"
        }
        val file = File(context.cacheDir, "sakha-${UUID.randomUUID()}.$ext")
        FileOutputStream(file).use { it.write(bytes) }
        return file
    }

    private suspend fun playFile(file: File): Long {
        val done = CompletableDeferred<Long>()
        val started = System.currentTimeMillis()
        try {
            val player = MediaPlayer().apply {
                setAudioAttributes(
                    AudioAttributes.Builder()
                        .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                        .setUsage(AudioAttributes.USAGE_ASSISTANT)
                        .build()
                )
                setDataSource(context, Uri.fromFile(file))
                setOnCompletionListener {
                    done.complete(System.currentTimeMillis() - started)
                }
                setOnErrorListener { _, what, extra ->
                    done.complete(0L)
                    if (config.debugMode) Log.w(TAG, "MediaPlayer error: $what/$extra")
                    true
                }
                prepare()
                start()
            }
            currentPlayer = player
            val ms = done.await()
            try { player.release() } catch (_: Exception) {}
            currentPlayer = null
            return ms
        } catch (e: Exception) {
            if (config.debugMode) Log.w(TAG, "Playback failed: ${e.message}")
            return 0L
        }
    }

    // ------------------------------------------------------------------------
    // System TTS fallback
    // ------------------------------------------------------------------------

    private fun ensureFallbackTts() {
        if (fallbackTts != null) return
        fallbackTts = TextToSpeech(context.applicationContext) { status ->
            if (status == TextToSpeech.SUCCESS) {
                val res = fallbackTts?.setLanguage(config.language.sttLocale)
                fallbackTtsReady = (res != TextToSpeech.LANG_MISSING_DATA &&
                        res != TextToSpeech.LANG_NOT_SUPPORTED)
            } else {
                fallbackTtsReady = false
            }
        }
    }

    private suspend fun speakWithFallback(text: String): Long {
        val tts = fallbackTts ?: return 0L
        val started = System.currentTimeMillis()
        val done = CompletableDeferred<Unit>()
        val utteranceId = UUID.randomUUID().toString()
        tts.setOnUtteranceProgressListener(object : UtteranceProgressListener() {
            override fun onStart(id: String?) {}
            override fun onDone(id: String?) { done.complete(Unit) }
            @Deprecated("Deprecated in Java")
            override fun onError(id: String?) { done.complete(Unit) }
            override fun onError(id: String?, errorCode: Int) { done.complete(Unit) }
        })
        tts.speak(text, TextToSpeech.QUEUE_ADD, null, utteranceId)
        done.await()
        return System.currentTimeMillis() - started
    }

    // ------------------------------------------------------------------------
    // Internal job sealed
    // ------------------------------------------------------------------------

    private sealed class PlayJob {
        data class Synthesise(val text: String, val isSanskrit: Boolean) : PlayJob()
        data class Pause(val durationMs: Long) : PlayJob()
        object Filter : PlayJob()
    }
}

File: kiaanverse-mobile/apps/mobile/native/ios/Sources/Sakha/SakhaTtsPlayer.swift
/**
 * Sakha TTS Player (iOS port)
 *
 * Direct port of apps/mobile/native/android/.../SakhaTtsPlayer.kt.
 * Sequential, pause-aware playback queue for the Sakha voice. Two
 * backends:
 *
 *   1. Sarvam REST   — primary, prose + Sanskrit verses.
 *                      POST {backend}/api/voice-companion/synthesize
 *                      → audio bytes (base64 in JSON or raw audio)
 *                      → AVAudioPlayer playback.
 *   2. AVSpeech…    — fallback when Sarvam fails AND the language has
 *                      a system voice. NEVER used for Sanskrit (the
 *                      system voice mangles it).
 *
 * Concurrency model:
 *   - One Task pulls events off `incoming` (AsyncStream<PauseEvent>).
 *   - One Task drains the synthesis queue, plays each clip back-to-back,
 *     and signals progress via the listener.
 *
 * This pipelines synthesis ahead of playback (one sentence plays while
 * the next is being synthesised) without overlapping audio — Sakha
 * pacing depends on ordered, gapless playback.
 *
 * Threading:
 *   - URLSession runs on its own queue.
 *   - AVAudioPlayer.play() is main-thread safe.
 *   - Listener callbacks fire on .main (mirrors Kotlin Dispatchers.Main).
 */

import Foundation
import AVFoundation

public final class SakhaTtsPlayer: NSObject {

    private let config: SakhaVoiceConfig
    private weak var listener: (any SakhaVoiceListener)?
    private let session: URLSession

    /// Latest PauseEvent stream — recreated on each `start()` so a session
    /// that stops + restarts (e.g. barge-in) gets fresh continuations.
    private var incomingContinuation: AsyncStream<PauseEvent>.Continuation?
    private var ingestTask: Task<Void, Never>?
    private var playbackTask: Task<Void, Never>?

    private var fallbackSynth: AVSpeechSynthesizer?

    private var currentPlayer: AVAudioPlayer?
    private var totalSpokenMs: Int64 = 0
    private var pauseCount: Int = 0
    private var stopped: Bool = false

    /// Provides a bearer token for synthesis requests (mirrors the SSE client).
    private let authTokenProvider: () async -> String?

    public init(
        config: SakhaVoiceConfig,
        listener: any SakhaVoiceListener,
        authTokenProvider: @escaping () async -> String?
    ) {
        self.config = config
        self.listener = listener
        self.authTokenProvider = authTokenProvider

        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 8
        cfg.timeoutIntervalForResource = 20
        self.session = URLSession(configuration: cfg)
        super.init()
    }

    // MARK: - Public lifecycle

    /// Start a new session. Call `enqueue(...)` to feed events and
    /// `stop()` to halt + free resources.
    public func start() {
        stopped = false
        totalSpokenMs = 0
        pauseCount = 0

        // Build a fresh AsyncStream for this session.
        let (stream, continuation) = AsyncStream<PauseEvent>.makeStream(bufferingPolicy: .unbounded)
        self.incomingContinuation = continuation

        ingestTask?.cancel()
        playbackTask?.cancel()

        // Single sequential consumer — no separate ingest queue needed
        // on iOS because synthesis + playback run in the same task and
        // structured concurrency naturally pipelines: while one segment
        // is playing on AVAudioPlayer, the next can be `await`-synthesised
        // by an inner Task. (Simpler than the Kotlin two-channel design;
        // gapless behavior is preserved.)
        playbackTask = Task { [weak self] in
            guard let self = self else { return }
            for await event in stream {
                if self.stopped { break }
                await self.consume(event)
            }
        }
    }

    public func enqueue(_ event: PauseEvent) {
        incomingContinuation?.yield(event)
    }

    /// Mark end-of-stream — drains the queue, then tasks complete.
    public func finish() {
        incomingContinuation?.finish()
    }

    /// Halt playback and synthesis; safe to call from any thread.
    public func stop() {
        stopped = true
        ingestTask?.cancel()
        playbackTask?.cancel()
        incomingContinuation?.finish()
        currentPlayer?.stop()
        currentPlayer = nil
        fallbackSynth?.stopSpeaking(at: .immediate)
    }

    public func shutdown() {
        stop()
        fallbackSynth?.delegate = nil
        fallbackSynth = nil
    }

    public func statsTotalSpokenMs() -> Int64 { totalSpokenMs }
    public func statsPauseCount() -> Int { pauseCount }

    // MARK: - Event consumption

    private func consume(_ event: PauseEvent) async {
        if stopped { return }
        switch event {
        case .speak(let text, let isSanskrit):
            await speak(text: text, isSanskrit: isSanskrit)
        case .pause(let durationMs):
            pauseCount += 1
            await emitOnMain { [listener] in
                listener?.sakhaVoice(pause: durationMs)
            }
            try? await Task.sleep(nanoseconds: UInt64(max(durationMs, 0)) * 1_000_000)
        case .filter:
            await emitOnMain { [listener] in
                listener?.sakhaVoiceFilterFail()
            }
        }
    }

    // MARK: - Synthesis + playback

    private func speak(text: String, isSanskrit: Bool) async {
        if stopped { return }

        // Persona guard runs in the manager BEFORE enqueuing — we don't
        // re-run it here; trust upstream.
        let voiceId = isSanskrit ? config.sanskritVoiceId : config.sakhaVoiceId

        let started = nowMs()
        let audioData: Data? = await synthesize(text: text, voiceId: voiceId, language: config.language)

        if stopped { return }

        if let data = audioData, !data.isEmpty {
            await playAudio(data: data)
        } else if !isSanskrit {
            // Fallback to AVSpeechSynthesizer — never for Sanskrit.
            await speakWithFallback(text: text)
        } else {
            await emitOnMain { [listener] in
                listener?.sakhaVoice(error: .ttsError("synthesis failed (sanskrit, no fallback)"))
            }
        }

        if !stopped {
            totalSpokenMs += max(nowMs() - started, 0)
            await emitOnMain { [listener] in
                listener?.sakhaVoice(spokenSegment: text, isSanskrit: isSanskrit)
            }
        }
    }

    private func synthesize(text: String, voiceId: String, language: SakhaLanguage) async -> Data? {
        let url = "\(trimSlash(config.backendBaseUrl))\(config.voiceCompanionPathPrefix)/synthesize"
        guard let endpoint = URL(string: url) else { return nil }

        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
        request.setValue("kiaanverse-ios-sakha-voice", forHTTPHeaderField: "X-Client")
        if let token = await authTokenProvider(), !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        let payload: [String: Any] = [
            "text": text,
            "voice_id": voiceId,
            "language": language.wire,
        ]
        request.httpBody = try? JSONSerialization.data(withJSONObject: payload, options: [])

        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse, http.statusCode < 400 else {
                return nil
            }
            // Two response shapes per Android: raw audio bytes OR JSON
            // with base64 audio under "audio" / "data".
            if http.value(forHTTPHeaderField: "Content-Type")?.contains("application/json") == true {
                if let json = try? JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] {
                    let b64 = (json["audio"] as? String) ?? (json["data"] as? String) ?? ""
                    if !b64.isEmpty {
                        return Data(base64Encoded: b64)
                    }
                }
                return nil
            }
            return data
        } catch {
            return nil
        }
    }

    private func playAudio(data: Data) async {
        await MainActor.run {
            do {
                let player = try AVAudioPlayer(data: data)
                player.delegate = self
                self.currentPlayer = player
                player.prepareToPlay()
                player.play()
            } catch {
                self.currentPlayer = nil
            }
        }
        // Wait for playback to finish (or stop() to nil out the player).
        while !stopped, let p = currentPlayer, p.isPlaying {
            try? await Task.sleep(nanoseconds: 40_000_000) // 40ms tick
        }
    }

    private func speakWithFallback(text: String) async {
        await MainActor.run {
            if self.fallbackSynth == nil {
                self.fallbackSynth = AVSpeechSynthesizer()
            }
            let utterance = AVSpeechUtterance(string: text)
            utterance.voice = AVSpeechSynthesisVoice(language: self.config.language.sttLocaleId)
            utterance.rate = AVSpeechUtteranceDefaultSpeechRate * 0.95
            self.fallbackSynth?.speak(utterance)
        }
        // Poll until the synthesizer finishes (no clean async API on iOS 17).
        while !stopped, fallbackSynth?.isSpeaking == true {
            try? await Task.sleep(nanoseconds: 40_000_000)
        }
    }

    // MARK: - Helpers

    private func nowMs() -> Int64 {
        return Int64(Date().timeIntervalSince1970 * 1000.0)
    }

    private func trimSlash(_ s: String) -> String {
        if s.hasSuffix("/") { return String(s.dropLast()) }
        return s
    }

    private func emitOnMain(_ block: @escaping () -> Void) async {
        await MainActor.run { block() }
    }
}

// MARK: - AVAudioPlayerDelegate

extension SakhaTtsPlayer: AVAudioPlayerDelegate {
    public func audioPlayerDidFinishPlaying(_ player: AVAudioPlayer, successfully: Bool) {
        if currentPlayer === player {
            currentPlayer = nil
        }
    }
}
